universal/error-handling.md

```diff
@@ -0,0 +1,149 @@
+# Universal / Error handling
+
+An error can be categorized by many criteria, whether
+- it's **critical software** or it's **non-critical software**
+- it's **user facing** or **not user facing**
+- there **is a fallback** ir **there is no fallback**
+- it's on the **frontend** or **backend**
+- the code is **synchronous** or **asynchronous**
+- a certain language or library is used
+
+These are general rules aimed to help making a decision about how to handle certain errors. For errors texts refer to the [wording guideline](https://github.com/c-hive/guides/issues/2).
+
+## Critical software
+
+There's always a fallback strategy. Errors are always reported. Possible bugs are considered for graceful error handling.
+
+- Example: bank system
+
+## Non-critical software
+
+Bugs are not considered for graceful error handling. There isn't always a fallback strategy but errors should be reported.
+
+## User facing error
+
+Often happens on the frontend.
+
+#### By default, the user must be notified and a fallback strategy must be implemented.
+
+- Example: a form validation failed on post. The validation error is shown, the field is higlighted and the previous value may or may not be removed.
+
+- Example: worst case, some unexpected error happens. A generic error is shown: "Something unexpected happened, save your changes offline and [reload the page]()."
+
+#### In certain (specifically identified) cases the user may not be notified.
+
+- Example: an async fethcing keeps the website up to date with other changes but it fails. This may fail silently in the bakckground (but must be handled explicitly). Fallback strategy could be retrying a few times and, if it keeps failing, showing a generic error to the user: "The page in running with limited functionality, [reloading the page]() might help resolve it."
+
+#### Expose the minimal necessary information to the user about the error.
+
+- Example: GOOD: "Something unexpected happened", BAD: "JSON.parseError"
+- Example: GOOD: "The page in running with limited functionality", BAD: "XYZ feature is not available"
+
+## Not user facing error
+
+Often happens on the backend. In certain cases a generic fallback strategy should be implemented (e.g. retry).
+
+#### Fail hard unless there's a fallback strategy
+
+GOOD
+
+```
+thisMightThrow();
```

Comment:
This means that the right approach - in this case - is to let the app or whatever fail and figure out a fallback strategy, right? Instead of logging the error to the console as it wouldn't give an appropriate feedback - and the error is also caught, thus the app doesn't crash. Probably that's why it's called *fail hard*. :)

Reply:
Yes, just let an exception happen unless you know how to handle it.